ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -23,123 +29,204 @@ import           Network.Socket hiding (recv, recvFrom, send, sendTo)
+          if BL.null buf
+              then throwM $ Mx.MuxError Mx.MuxBearerClosed "Socket closed when reading data" callStack
+              else recvLen' (l - fromIntegral (BL.length buf)) (buf : bufs)
+
+      writeSocket :: Mx.MuxSDU ptcl -> IO (Time IO)
+      writeSocket sdu = do
+          --say "write"
+          ts <- getMonotonicTime
+          let sdu' = sdu { Mx.msTimestamp = Mx.RemoteClockModel $ fromIntegral $ ts .&. 0xffffffff }
+              buf = Mx.encodeMuxSDU sdu'
+          --hexDump buf ""
+          Socket.sendAll sd buf
+          return ts
+
+      closeSocket :: IO ()
+      closeSocket = close sd
+
+      sduSize :: IO Word16
+      sduSize = do
+          -- XXX it is really not acceptable to call getSocketOption for every SDU we want to send
+          mss <- getSocketOption sd MaxSegment
+          -- 1260 = IPv6 min MTU minus TCP header, 8 = mux header size
+          return $ fromIntegral $ max (1260 - 8) (min 0xffff (15 * mss - 8))
+
 
 hexDump :: BL.ByteString -> String -> IO ()
 hexDump buf out | BL.empty == buf = say out
 hexDump buf out = hexDump (BL.tail buf) (out ++ printf "0x%02x " (BL.head buf))
 
+-- |
+-- Like @'runNetworkNodeWithSocket'@ but it allows to run an action when an
+-- exception is raised by the mux layer.  This is useful in tests.
+--
+runNetworkNodeWithSocket'
+  :: forall ptcl.
+     ( Mx.ProtocolEnum ptcl
+     , Ord ptcl
+     , Enum ptcl
+     , Bounded ptcl
+     )
+  => NetworkInterface ptcl AddrInfo IO
+  -> Maybe (Maybe SomeException -> IO ())
+  -> IO (NetworkNode AddrInfo IO)
+runNetworkNodeWithSocket' NetworkInterface {nodeAddress, knownVersions, protocols} k = do
+      (sd, hdl) <- startNode
+      return $ NetworkNode {
+          connectTo,
+          killNode  = closeConn sd hdl
+        }
+    where
+      mpds :: Mx.SomeVersion -> Maybe (Mx.MiniProtocolDescriptions ptcl IO)
+      mpds sv = (miniProtocolDescription . ) <$> protocols sv
+
+      startNode :: IO (Socket, Async ())
+      startNode =
+          bracketOnError
+            (socket (addrFamily nodeAddress) Stream defaultProtocol)
+            close
+            (\sd -> do
+                setSocketOption sd ReuseAddr 1
+                setSocketOption sd ReusePort 1
+                bind sd (addrAddress nodeAddress)
+                listen sd 2
+                rh <- async (server sd)
+                return (sd, rh)
+            )
+        where
+          server sd = forever $ do
+            (client, _) <- accept sd
+            aid <- async $ larval client
+            void $ async $ watcher client aid
+
+          larval sd = do
+            bearer <- socketAsMuxBearer sd
+            Mx.muxBearerSetState bearer Mx.Connected
+            Mx.muxStart knownVersions mpds bearer Mx.StyleServer k
+
+          watcher sd aid = do
+            res_e <- waitCatch aid
+            case res_e of
+              Left  e -> do
+                close sd
+                sequence_ $ k <*> (Just (Just e))
```

Comment:
This is very opaque. What's going on here?

Reply:
Oh this is the exception callback

Reply:
And the exception callback will go away, when we'll add proper logging to mux.